File: apps/ios/Packages/StoryArcKit/Sources/Formats/AdoptingArchive.swift
public import Foundation

import Synchronization

/// An archive that can change where its bytes come from without the reader noticing.
///
/// `offline-downloads`' *Reading while downloading* asks a publication opened by streaming to
/// "switch to the local copy when the download completes, without interrupting reading". This
/// is the switch. Every page the reader has already decoded stays decoded, the page list is
/// never rewritten, and no position is read or written here — so a reader on page 14 is on
/// page 14 before and after, and nothing reopens under them.
///
/// The reader wraps whatever it opened, local or streamed, so there is one shape above this
/// line rather than two. Android's `AdoptingArchive` is the same object.
///
/// A class holding a lock rather than a value, for both halves of what this has to be:
/// ``ComicArchiveReading`` is `Sendable` and decoding runs off the main actor, so the source
/// can be read from anywhere — and a switch has to be seen by every holder of the reference
/// at once, which is exactly what a value type cannot do.
public final class AdoptingArchive: ComicArchiveReading {
    private let state: Mutex<any ComicArchiveReading>

    public init(_ reading: any ComicArchiveReading) {
        state = Mutex(reading)
    }

    public var pages: [PageEntry] { state.withLock { $0.pages } }

    public var skippedPageCount: Int { state.withLock { $0.skippedPageCount } }

    public var coverPage: PageEntry? { state.withLock { $0.coverPage } }

    public var doublePageIndices: [Int] { state.withLock { $0.doublePageIndices } }

    public func data(for page: PageEntry) async throws -> Data {
        // The lock is released before the read: a page fetched over a range request takes as
        // long as the network does, and holding a mutex across it would stop the reader.
        try await state.withLock { $0 }.data(for: page)
    }

    /// Reads from `local` from now on, when it holds the same pages.
    ///
    /// Refused when it does not, and that is the whole of the safety here: the reader indexes
    /// its pages by position, so an archive listing different entries would move every page
    /// the reader has not decoded yet. A refusal leaves this archive exactly as it was.
    ///
    /// - Returns: whether the switch happened.
    @discardableResult
    public func adopt(_ local: any ComicArchiveReading) -> Bool {
        state.withLock { reading in
            guard local.pages == reading.pages else { return false }
            reading = local
            return true
        }
    }
}

File: apps/android/core/format/src/main/kotlin/app/storyarc/core/format/AdoptingArchive.kt
package app.storyarc.core.format

/**
 * An archive that can change where its bytes come from without the reader noticing.
 *
 * `offline-downloads`' *Reading while downloading* asks a publication opened by streaming to
 * "switch to the local copy when the download completes, without interrupting reading". This
 * is the switch. Every page the reader has already decoded stays decoded, the page list is
 * never rewritten, and no position is read or written here -- so a reader on page 14 is on
 * page 14 before and after, and nothing reopens under them.
 *
 * The reader wraps whatever it opened, local or streamed, so there is one shape above this
 * line rather than two. iOS's `AdoptingArchive` is the same object.
 */
class AdoptingArchive(private var reading: ComicArchiveReading) : ComicArchiveReading {

    /**
     * What [reading] took over from, held until this archive closes.
     *
     * Not closed at the moment of the swap: a page decode may be reading from it on another
     * thread, and closing a source under a read turns an invisible switch into a broken page.
     * Both are released together by [close], which the reader calls when it is done.
     */
    private var replaced: ComicArchiveReading? = null

    override val pages: List<PageEntry> get() = reading.pages

    override val skippedPageCount: Int get() = reading.skippedPageCount

    override val coverPage: PageEntry? get() = reading.coverPage

    override val doublePageIndices: List<Int> get() = reading.doublePageIndices

    override suspend fun data(page: PageEntry): ByteArray = reading.data(page)

    /**
     * Reads from [local] from now on, when it holds the same pages.
     *
     * Refused when it does not, and that is the whole of the safety here: the reader indexes
     * its pages by position, so an archive listing different entries would move every page
     * the reader has not decoded yet. A refusal leaves this archive exactly as it was and
     * closes [local], which the caller never held.
     *
     * @return whether the switch happened.
     */
    fun adopt(local: ComicArchiveReading): Boolean {
        if (local.pages != reading.pages) {
            local.close()
            return false
        }
        replaced = reading
        reading = local
        return true
    }

    override fun close() {
        reading.close()
        replaced?.close()
        replaced = null
    }
}
